### PR TITLE
build: bottlerocket-test-tools

### LINKED ISSUE #335: build: go builds are risky
We can see stuff like this which we have no control over:

```
=> CANCELED [sonobuoy-build 5/7] RUN go mod vendor                                                                          40.7s
 => ERROR [aws-iam-authenticator-build 5/6] RUN go mod vendor                                                                40.1s
 => [build-src 5/5] RUN cargo install boringtun     --target x86_64-bottlerocket-linux-musl     --root .                     24.4s
------                                                                                                                             
 > [aws-iam-authenticator-build 5/6] RUN go mod vendor:                                                                            
#17 39.87 go: k8s.io/client-go@v0.16.8 requires                                                                                    
#17 39.87       cloud.google.com/go@v0.38.0: unrecognized import path "cloud.google.com/go": reading https://cloud.google.com/go?go-get=1: 500 Internal Server Error                
```

I've seen problems with the Go builds many times vs. basically never with cargo builds.

We need to get the Go builds out of our critical path for iterating on and using TestSys

### PR DESCRIPTION
**Pull request by webern**
Move the third party test binaries to their own container image build.

**Issue number:**

Fixes #335

**Description of changes:**

We separate concerns by moving the Go-tools builds to their own `Dockerfile`.
This image will be published to ECR public so that we do not have to build these tools and extract their licenses each time we build our agent containers.

**Testing done:**

- I created an ECR Public repo with an obscure name.
- Set `TOOLS_IMAGE=myobscure.com/whatever/tools-test:v0.0.1-alpha`
- I used `make tools -e TOOLS_IMAGE=${TOOLS_IMAGE}`
- I pushed this image to $TOOLS_IMAGE
- Then I purged docker of all images and state
- Then I built all the testing containers
  - `make ec2-resource-agent -e TOOLS_IMAGE=${TOOLS_IMAGE}`
  - `make ecs-resource-agent -e TOOLS_IMAGE=${TOOLS_IMAGE}`
  - `make eks-resource-agent -e TOOLS_IMAGE=${TOOLS_IMAGE}`
  - `make sonobuoy-test-agent -e TOOLS_IMAGE=${TOOLS_IMAGE}`
  - `make ecs-test-agent -e TOOLS_IMAGE=${TOOLS_IMAGE}`
- I used these images to run a sonobuoy quick and an ecs smoke test.

It worked!

Also CI shows that you can still use a local image without pushing, you just have to build it before everything else.
If you don't feel like building this stuff though, you can just use the default and it will be pulled from ECR Public (once published)!

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

### CODE REVIEW

**Comment by webern** [comment]
^ rebase

**Comment by webern** [comment]
^ rebase

**Comment by webern** [comment]
^ rebase

**Comment by webern** [comment]
^ add a changelog file for the tools directory

**Comment by webern** [comment]
^ push the date in the changelog

File: aws-iam-authenticator_0.5.3.tar.gz [hunk 1]
```diff
@@ -0,0 +1,2 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (etungsten) [inline aws-iam-authenticator_0.5.3.tar.gz]: This file doesn't look like it should be here. Failed curl?

File: tools/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+    --root /home/builder
+
+# =^..^= =^..^= =^..^= =^..^= =^..^= =^..^= =^..^= =^..^= =^..^= =^..^= =^..^= =^..^= =^..^= =^..^= =^..^= =^..^= =^..^=
+FROM build-go as eksctl-build
+
+USER root
+RUN mkdir -p /usr/share/licenses/eksctl && \
+    chown -R builder:builder /usr/share/licenses/eksctl
+
+ARG EKSCTL_VERSION=0.82.0
```
Comment (etungsten) [inline tools/Dockerfile]: non-blocking: before pushing to public ECR, please bump this to the latest version.

File: tools/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+    install -m 0755 kubeadm_${K8S_VERSION}_${GOOS}_${GOARCH} /tmp/kubeadm
+
+# =^..^= =^..^= =^..^= =^..^= =^..^= =^..^= =^..^= =^..^= =^..^= =^..^= =^..^= =^..^= =^..^= =^..^= =^..^= =^..^= =^..^=
+FROM build-go as sonobuoy-build
+
+USER root
+RUN mkdir -p /usr/share/licenses/sonobuoy && \
+    chown -R builder:builder /usr/share/licenses/sonobuoy
+
+ARG SONOBUOY_VERSION=0.53.2
```
Comment (etungsten) [inline tools/Dockerfile]: non-blocking: please bump this to the latest version before cutting a release of the tools container image.

**Review by etungsten — CHANGES_REQUESTED** [review]

**Comment by webern** [comment]
^ updated versions of eksctl and sonobuoy, but the go builds broke! working on it.

**Comment by webern** [comment]
^ gitignore tar files and attempt to fix the eksctl build

File: tools/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+WORKDIR /home/builder/eksctl/
+# TODO - remove this workaround for a bad go.sum hash
+# caused by https://github.com/blizzy78/varnamelen/issues/13#issuecomment-1079040934
+RUN \
+    # this sed is compatible with BSD and GNU versions of sed. it removes the version requirement in go.mod
+    sed -i.bak 's/go 1.18//' go.mod && \
+    # we need to delete and re-create the go.sum file to avoid a bad hash contained in it
+    rm go.sum && \
+    go clean -modcache && \
+    go mod tidy
```
Comment (jpculp) [inline tools/Dockerfile]:
```suggestion
RUN \
  sed -i 's,github.com/blizzy78/varnamelen v0.6.1 h1:kttPCLzXFa+0nt++Cw9fb7GrSSM4KkyIAoX/vXsbuqA=,github.com/blizzy78/varnamelen v0.6.1 h1:iYAU/3A6cpfRm2ZI0P/lece4jsc7GEbzsxTu+vBCChQ=,' go.sum ; \
  sed -i 's,github.com/blizzy78/varnamelen v0.6.1/go.mod h1:zy2Eic4qWqjrxa60jG34cfL0VXcSwzUrIx68eJPb4Q8=,github.com/blizzy78/varnamelen v0.6.1/go.mod h1:mGBHm+Uo4e8JnZEKHRoZgVEOQdSBdQfY/x+k4NAXBWA=,' go.sum
R
```
I would just fix the sums manually to keep things simple for now. I'm not sure if we'd rather make this a patch than a `sed` though.

**Review by jpculp — CHANGES_REQUESTED** [review]
You'll need to add the following to `tools/clarify.toml`:
```
[clarify."github.com/dave/dst"]
expression = "MIT AND BSD-3-Clause"
license-files = [
    { path = "LICENSE", hash = 0x8f288482},
]
```

File: tools/Dockerfile (hunk 4 above)
Comment (webern) [reply tools/Dockerfile]: This seems easier than a patch to me. Thank you I will try this.

**Comment by webern** [comment]
^ implement @jpculp's fix for the eksctl build

**Review by etungsten — APPROVED** [review]

**Review by jpculp — APPROVED** [review]

**Comment by webern** [comment]
^ move the date again.

Also, I have re-tested and the new versions of eksctl and sonobuoy are working.

**Pull request merged**